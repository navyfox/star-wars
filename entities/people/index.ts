export * from './people';

// export default slice.peopleSliceReducer;
