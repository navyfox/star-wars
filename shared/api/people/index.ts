export * from './people';
export * from './types';
