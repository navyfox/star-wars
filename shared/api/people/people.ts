import axios from 'axios';
import { PeopleApiResponse } from './types';

import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import { HYDRATE } from 'next-redux-wrapper';

export const peopleApi = createApi({
  baseQuery: fetchBaseQuery({
    baseUrl: `${process.env.NEXT_PUBLIC_STAR_WARS_API}/`,
  }),
  extractRehydrationInfo(action, { reducerPath }) {
    if (action.type === HYDRATE) {
      return action.payload[reducerPath];
    }
  },
  tagTypes: [],
  endpoints: (builder) => ({
    getPeopleByPage: builder.query<PeopleApiResponse, number>({
      query: (page) => `people/?page=${page}`,
      keepUnusedDataFor: 120,
    }),
  }),
});

export const {
  useGetPeopleByPageQuery,
  util: { getRunningQueriesThunk },
} = peopleApi;

export const { getPeopleByPage } = peopleApi.endpoints;

export const getPersonByID = async (id: number | string) =>
  axios.get(`${process.env.NEXT_PUBLIC_STAR_WARS_API}/people/${id}`, {
    headers: {
      'Content-Type': 'application/json',
    },
  });
