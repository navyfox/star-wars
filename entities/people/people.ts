import { createSlice, createAction, createAsyncThunk } from '@reduxjs/toolkit';

import { HYDRATE } from 'next-redux-wrapper';
import { Person } from '@/shared/api/people';

// mock
const savePersonApi = (person: Person | null) =>
  new Promise((resolve, reject) => {
    setTimeout(() => {
      console.log('SAVE_PERSON'), person;
      resolve({ isSaving: true });
    }, 1000);
  });

export interface PeopleState {
  person: Person | null;
  isSaving: boolean;
  isPending: boolean;
  error: string | null;
}

const initialState: PeopleState = {
  person: null,
  isSaving: true,
  isPending: false,
  error: null,
};

const hydrate = createAction<{ people: PeopleState }>(HYDRATE);

export const savePerson = createAsyncThunk(
  'people/savePerson',
  async (person: Person | null, { rejectWithValue }) => {
    try {
      const response = await savePersonApi(person);
      return response;
    } catch (error: any) {
      return rejectWithValue(error.response.data);
    }
  }
);

export const peopleSlice = createSlice({
  name: 'people',
  initialState,
  reducers: {
    setPerson(state, action) {
      state.person = { ...state.person, ...action.payload };
    },
    setIsSaving(state, action) {
      state.isSaving = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(hydrate, (state, action) => {
      return {
        ...state,
        ...action.payload.people,
      };
    });
    builder.addCase(savePerson.pending, (state) => {
      state.isPending = true;
    });
    builder.addCase(savePerson.fulfilled, (state, action) => {
      state.isSaving = (action.payload as { isSaving: boolean }).isSaving;
    });
    builder.addCase(savePerson.rejected, (state, action) => {
      state.isSaving = false;
      state.error = action.payload as string;
    });
  },
});

export const { setPerson, setIsSaving } = peopleSlice.actions;

export default peopleSlice.reducer;
