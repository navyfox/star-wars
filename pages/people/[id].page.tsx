import React, { ChangeEvent } from 'react';
import Head from 'next/head';
import { GetServerSideProps } from 'next';
import Container from '@mui/material/Container';
import { useSelector, useDispatch } from 'react-redux';
import { setPerson } from '@/entities/people';
import { wrapperStore } from '@/settings/store';
import Box from '@mui/material/Box';
import PersonEditCard from '@/widgets/person-edit-card';
import { getPersonByID } from '@/shared/api/people';
import { Person } from '@/shared/api/people';
import { savePerson, setIsSaving } from '@/entities/people';
import Button from '@mui/material/Button';
import WrapperCard from '@/shared/ui/wrapper-card';
import Stack from '@mui/material/Stack';
import type { NextPage } from 'next/types';

export const PersonPage: NextPage = () => {
  const dispatch = useDispatch<AppThunk>();
  const person = useSelector((state: AppState) => state.people.person);

  const onSave = () => {
    dispatch(savePerson(person));
  };

  const onChangePerson = (event: ChangeEvent<HTMLTextAreaElement | HTMLInputElement>) => {
    dispatch(
      setPerson({
        [event.target.id]: event.target.value,
      })
    );
    dispatch(setIsSaving(false));
  };
  return (
    <>
      <Head>
        <title>People from Star Wars</title>
        <meta name="viewport" content="width=device-width, initial-scale=1" />
      </Head>
      <Container maxWidth="sm">
        <Box marginTop={5} data-testid="person-card">
          <WrapperCard>
            <PersonEditCard person={person} onChangePerson={onChangePerson} />
            <Stack marginTop={5} marginBottom={1} alignItems="center">
              <Button variant="outlined" onClick={onSave}>
                SAVE
              </Button>
            </Stack>
          </WrapperCard>
        </Box>
      </Container>
    </>
  );
};

export const getServerSideProps: GetServerSideProps = wrapperStore.getServerSideProps(
  (store) =>
    async ({ query: { id = '1' } }) => {
      const pageIndex = Array.isArray(id) ? parseInt(id[0]) : parseInt(id) || 1;
      const person = await getPersonByID(pageIndex).then((resp) => resp.data as Person);
      await store.dispatch(setPerson(person));
      return { props: {} };
    }
);

export default PersonPage;
