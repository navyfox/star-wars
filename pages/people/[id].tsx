import React, { ChangeEvent } from 'react';
import Head from 'next/head';
import { GetServerSideProps } from 'next';
import Container from '@mui/material/Container';
import { useSelector, useDispatch } from 'react-redux';
import { setPerson } from '@/entities/people';
import { wrapperStore } from '@/app/store';
import Box from '@mui/material/Box';
import PersonEditCard from '@/widgets/person-edit-card';
import { getPersonByID } from '@/shared/api/people';
import { Person } from '@/shared/api/people';

export const PersonPage = () => {
  const dispatch = useDispatch();
  const person = useSelector((state: AppState) => state.people.person);

  const onChangePerson = (event: ChangeEvent<HTMLTextAreaElement | HTMLInputElement>) => {
    dispatch(
      setPerson({
        [event.target.id]: event.target.value,
      })
    );
  };
  return (
    <>
      <Head>
        <title>People from Star Wars</title>
        <meta name="viewport" content="width=device-width, initial-scale=1" />
      </Head>
      <Container maxWidth="sm">
        <Box marginTop={5}>
          <PersonEditCard person={person} onChangePerson={onChangePerson} />
        </Box>
      </Container>
    </>
  );
};

export const getServerSideProps: GetServerSideProps = wrapperStore.getServerSideProps(
  (store) =>
    async ({ query: { id = '1' } }) => {
      const pageIndex = Array.isArray(id) ? parseInt(id[0]) : parseInt(id) || 1;
      const person = await getPersonByID(pageIndex).then((resp) => resp.data as Person);
      await store.dispatch(setPerson(person));
      return { props: {} };
    }
);

export default PersonPage;
