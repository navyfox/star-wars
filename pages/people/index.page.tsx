import React from 'react';
import type { NextPage } from 'next/types';
import Head from 'next/head';
import { GetServerSideProps } from 'next';
import Container from '@mui/material/Container';
import { getPeopleByPage } from '@/shared/api/people';
import { Person } from '@/shared/api/people';
import PeoplePagination from '@/widgets/people-pagination';

interface PeoplePageProps {
  count: number;
  people: Person[];
  page: number;
}

export const PeoplePage: NextPage<PeoplePageProps> = ({ count, people, page }: PeoplePageProps) => {
  return (
    <>
      <Head>
        <title>People from Star Wars</title>
        <meta name="viewport" content="width=device-width, initial-scale=1" />
      </Head>
      <Container maxWidth="sm">
        <PeoplePagination count={count} people={people} page={page} />
      </Container>
    </>
  );
};

export const getServerSideProps: GetServerSideProps = async ({ query: { page = '1' } }) => {
  const pageIndex = Array.isArray(page) ? parseInt(page[0]) : parseInt(page) || 1;
  const data = await getPeopleByPage(pageIndex);

  return { props: { people: data.results, count: data.count, page: pageIndex } };
};

export default PeoplePage;
