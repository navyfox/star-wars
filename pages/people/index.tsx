import React from 'react';
import Head from 'next/head';
import { GetServerSideProps } from 'next';
import Container from '@mui/material/Container';
import { getPeopleByPage } from '@/shared/api/people';
import { Person } from '@/shared/api/people';
import PeoplePagination from '@/widgets/people-pagination';

export const PeoplePage = ({
  count,
  people,
  page,
}: {
  count: number;
  people: Person[];
  page: number;
}) => {
  return (
    <>
      <Head>
        <title>People from Star Wars</title>
        <meta name="viewport" content="width=device-width, initial-scale=1" />
      </Head>
      <Container maxWidth="sm">
        <PeoplePagination count={count} people={people} page={page} />
      </Container>
    </>
  );
};

export const getServerSideProps: GetServerSideProps = async ({ query: { page = '1' } }) => {
  const pageIndex = Array.isArray(page) ? parseInt(page[0]) : parseInt(page) || 1;
  const data = await getPeopleByPage(pageIndex);

  return { props: { people: data.results, count: data.count, page: pageIndex } };
};

export default PeoplePage;
