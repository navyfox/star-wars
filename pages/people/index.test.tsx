import React from 'react';
import { render, RenderResult } from '@testing-library/react';
import { GetServerSidePropsContext } from 'next';
import { PeoplePage, getServerSideProps } from './index.page';
import { getPeopleByPage, PeopleApiResponse } from '@/shared/api/people';
import renderer from 'react-test-renderer';

jest.mock('@/shared/api/people', () => ({
  getPeopleByPage: jest.fn(),
}));

describe('PeoplePage', () => {
  const props = {
    count: 0,
    people: [],
    page: 1,
  };

  let renderResult: RenderResult;

  beforeEach(() => {
    renderResult = render(<PeoplePage {...props} />);
  });

  it('renders correctly', () => {
    const component = renderer.create(<PeoplePage {...props} />).toJSON();
    expect(component).toMatchSnapshot();
  });

  it('should render without error', () => {
    const { container } = renderResult;
    expect(container).toBeInTheDocument();
  });
});

describe('getServerSideProps', () => {
  const mockApiResponse: PeopleApiResponse = {
    count: 10,
    results: [
      {
        name: 'Luke Skywalker',
        height: '172',
        mass: '77',
        hair_color: 'blond',
        skin_color: 'fair',
        eye_color: 'blue',
        birth_year: '19BBY',
        gender: 'male',
        url: 'https://swapi.dev/api/people/1/',
      },
    ],
  };

  beforeEach(() => {
    (getPeopleByPage as jest.Mock).mockResolvedValueOnce(mockApiResponse);
  });

  it('should return the expected props', async () => {
    const mockReq = {
      headers: {
        cookie: '',
      },
    } as any;

    const mockRes = {
      setHeader: jest.fn(),
      end: jest.fn(),
    } as any;

    const mockContext: GetServerSidePropsContext = {
      query: { page: '1' },
      req: mockReq,
      res: mockRes,
      resolvedUrl: '',
    };
    const { props } = (await getServerSideProps(mockContext)) as { props: Object };

    expect(getPeopleByPage).toHaveBeenCalledWith(1);
    expect(props).toHaveProperty('count', mockApiResponse.count);
    expect(props).toHaveProperty('people', mockApiResponse.results);
    expect(props).toHaveProperty('page', 1);
  });
});
