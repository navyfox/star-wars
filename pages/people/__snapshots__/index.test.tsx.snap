// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PeoplePage renders correctly 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthSm css-cuefkz-MuiContainer-root"
>
  <div
    className="MuiGrid2-root MuiGrid2-container MuiGrid2-direction-xs-row css-1ho1yxb-MuiGrid2-root"
    id="wrap"
  />
  <div
    className="MuiStack-root css-yzlqr0-MuiStack-root"
  >
    <nav
      aria-label="pagination navigation"
      className="MuiPagination-root MuiPagination-text css-1oj2twp-MuiPagination-root"
    >
      <ul
        className="MuiPagination-ul css-wjh20t-MuiPagination-ul"
      >
        <li>
          <a
            aria-disabled={true}
            aria-label="Go to previous page"
            className="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-disabled MuiPaginationItem-previousNext css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
            data-testid="pages"
            href="/people?page=0"
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
              data-testid="NavigateBeforeIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
              />
            </svg>
          </a>
        </li>
        <li>
          <a
            aria-disabled={true}
            aria-label="Go to next page"
            className="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-disabled MuiPaginationItem-previousNext css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
            data-testid="pages"
            href="/people?page=2"
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
              data-testid="NavigateNextIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              />
            </svg>
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;
